internal/consensus/reactor.go

```diff
@@ -229,34 +229,43 @@ func (r *Reactor) OnStop() {
 		r.state.Wait()
 	}
 
-	r.mtx.Lock()
-	peers := r.peers
-	r.mtx.Unlock()
-
-	// wait for all spawned peer goroutines to gracefully exit
-	for _, ps := range peers {
-		ps.closer.Close()
-	}
-	for _, ps := range peers {
-		ps.broadcastWG.Wait()
-	}
+	// Close closeCh to signal to all spawned goroutines to gracefully exit. All
+	// p2p Channels should execute Close().
+	close(r.closeCh)
 
 	// Close the StateChannel goroutine separately since it uses its own channel
 	// to signal closure.
 	close(r.stateCloseCh)
 	<-r.stateCh.Done()
 
-	// Close closeCh to signal to all spawned goroutines to gracefully exit. All
-	// p2p Channels should execute Close().
-	close(r.closeCh)
-
 	// Wait for all p2p Channels to be closed before returning. This ensures we
 	// can easily reason about synchronization of all p2p Channels and ensure no
 	// panics will occur.
 	<-r.voteSetBitsCh.Done()
 	<-r.dataCh.Done()
 	<-r.voteCh.Done()
 	<-r.peerUpdates.Done()
+
+	peers := make(map[types.NodeID]*PeerState)
+	r.mtx.Lock()
+
+	// Here, we make a copy of the map of peer states.
+	// The code below proceeds to access entries from the map outside of a lock.
+	// Goroutines launched by other functions may be deleting from the map but should
+	// no longer be appending to it since we waited for the peerUpdates channel to complete.
+	// Each of the methods below is safe to perform concurrently on the peers.
+	for id, state := range r.peers {
+		peers[id] = state
```

Comment:
are we worried about shared access to the contents of the  `PeerState` struct?

Reply:
No, the access that we take here is to methods that are covered by locks. Both `Wait()` and `Close()` use sync primitives.